File: coding/TripleKnock/htb/solver.cc
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

// Convert timestamp "DD/MM HH:MM" to total minutes since day 0
int total_minutes(const std::string& timestamp) {
    int day, month, hour, minute;
    sscanf(timestamp.c_str(), "%d/%d %d:%d", &day, &month, &hour, &minute);
    int days = (month - 1) * 30 + day - 1;
    return (days * 24 * 60) + (hour * 60) + minute;
}

int main() {
    int S, N;
    std::cin >> S >> N;
    std::cin.ignore(); // consume the newline after the first line

    std::unordered_map<std::string, std::vector<int>> user_times;

    for (int i = 0; i < S; ++i) {
        std::string line;
        std::getline(std::cin, line);

        if (line.find("[failure]") == std::string::npos) continue;

        size_t bracket_pos = line.find("[");
        std::string log_entry = line.substr(0, bracket_pos - 1); // trim trailing space
        std::istringstream iss(log_entry);
        std::string user, date, time;
        iss >> user >> date >> time;

        std::string timestamp = date + " " + time;
        user_times[user].push_back(total_minutes(timestamp));
    }

    std::vector<int> targeted_ids;

    for (auto& [user, times] : user_times) {
        std::sort(times.begin(), times.end());
        for (size_t i = 0; i + 2 < times.size(); ++i) {
            if (times[i + 2] - times[i] <= 10) {
                int id = std::stoi(user.substr(5)); // assumes user format "user_X"
                targeted_ids.push_back(id);
                break;
            }
        }
    }

    std::sort(targeted_ids.begin(), targeted_ids.end());

    for (size_t i = 0; i < targeted_ids.size(); ++i) {
        if (i > 0) std::cout << " ";
        std::cout << "user_" << targeted_ids[i];
    }
    std::cout << std::endl;

    return 0;
}

File: coding/TripleKnock/htb/solver.py
#!/usr/bin/env python3
from collections import defaultdict

def total_minutes(ts):
	# Parse day/month and hour:minute
	day_month, time = ts.split()
	day, month = map(int, day_month.split("/"))
	hour, minute = map(int, time.split(":"))
	# Total days since 01/01
	days = ((month-1) * 30) + day - 1  # -1 because 1/1 is day 0
	return (days * 24 * 60) + (hour * 60) + minute

def solve():
  S, N = map(int, input().split(' '))
  user_times = defaultdict(list)
  for _ in range(S):
    entry = input()
    if "failure" not in entry: continue
    entry = entry.split("[")[0].strip()
    user = entry.split(" ")[0]
    timestamp = " ".join(entry.split(" ")[1:])
    user_times[user].append(total_minutes(timestamp))

  targeted = []
  for user, times in user_times.items():
    times.sort()
    for i in range(len(times) - 2):
      if times[i + 2] - times[i] <= 10:
        targeted.append(int(user.replace("user_", "")))
        break

  targeted.sort()
  print(" ".join([f"user_{ID}" for ID in targeted]))

solve()